src/shared/clients/codecatalystClient.ts

```diff
@@ -676,3 +704,48 @@ class CodeCatalystClientInternal {
         return devenv
     }
 }
+
+/**
+ * Returns only the first-party repos from the given
+ * list of repository items.
+ */
+export async function getFirstPartyRepos(
+    client: CodeCatalystClient,
+    spaceName: CodeCatalystOrg['name'],
+    projectName: CodeCatalystProject['name'],
+    items?: Pick<ListSourceRepositoriesItem, 'name'>[]
+): Promise<ListSourceRepositoriesItems | undefined> {
+    if (items === undefined) {
+        return items
+    }
+
+    // Filter out any repos that are not 1P repos.
+    return (
+        await Promise.all(
+            items.map(async item => {
+                return (await isfirstPartyRepo(client, {
+                    spaceName,
+                    projectName,
+                    sourceRepositoryName: item.name,
+                }))
+                    ? item
+                    : undefined
+            })
+        )
+    ).filter(item => item !== undefined) as CodeCatalyst.ListSourceRepositoriesItem[]
+}
+
+/**
+ * Determines if a repo is first party (1P) vs third party (3P).
+ *
+ * 1P is Code Catalyst, 3P is something like Github.
+ */
+async function isfirstPartyRepo(
+    client: CodeCatalystClient,
+    codeCatalystRepo: GetSourceRepositoryCloneUrlsRequest
+): Promise<boolean> {
+    const url = await client.getRepoCloneUrl(codeCatalystRepo)
```

Comment:
👀  This is a O(N) performance cost for all customers, even if they have _zero_ third-party repos, right?